wasmtime/_memory.py

```diff
@@ -62,6 +64,64 @@ def data_ptr(self, store: Storelike) -> "ctypes._Pointer[c_ubyte]":
         """
         return ffi.wasmtime_memory_data(store._context, byref(self._memory))
 
+    def read(
+            self,
+            store: Storelike,
+            start: typing.Optional[int] = 0,
+            stop: typing.Optional[int] = None) -> bytearray:
+        """
+        Reads this memory starting from `start` and up to `stop`
+        and returns a copy of the contents as a byte array.
+
+        The indexing behavior of this method is similar to `list[start:stop]`
+        where negative starts can be used to read from the end, for example.
+        """
+        data_ptr = self.data_ptr(store)
+        size = self.data_len(store)
+        key = slice(start, stop, None)
+        start, stop, _ = key.indices(size)
+        val_size = stop - start
+        if val_size <= 0:
+            # return bytearray of size zero
+            return bytearray(0)
+        ptr_type = ctypes.c_ubyte * val_size
+        src_ptr = (ptr_type).from_address(ctypes.addressof(data_ptr.contents) + start)
+        return bytearray(src_ptr)
+
+    def write(
+            self,
+            store: Storelike,
+            value: typing.Any,
+            start: typing.Optional[int] = None,
+            stop: typing.Optional[int] = None) -> int:
```

Comment:
I think it would be best to remove the `stop` argument here. I understand you mentioned earlier that it can be used to reduce the size of the write but that can also be done by slicing the `value` argument here too. A downside of having the `stop` argument here is that this needs to document semantics of "what if the desired range is bigger than the `value` slice?" and I would prefer to not have to answer such a question.

Reply:
Additionally removing this `stop` argument removes the need to have `min`/etc logic and simplifies the indexing.